consensus/istanbul/backend/internal/enodes/val_enode_table.go

```diff
@@ -0,0 +1,145 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the celo library. If not, see <http://www.gnu.org/licenses/>.
+
+package enodes
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	// errOldAnnounceMessage is returned when the received announce message's block number is earlier
+	// than a previous received message
+	errOldAnnounceMessage = errors.New("old announce message")
+)
+
+// Entries for the valEnodeTable
+type validatorEnode struct {
+	enodeURL string
+	view     *istanbul.View
+}
+
+func (ve *validatorEnode) String() string {
+	return fmt.Sprintf("{enodeURL: %v, view: %v}", ve.enodeURL, ve.view)
+}
+
+// ValidatorEnodeTable represents a Map that can be accessed either
+// by address or enode
+type ValidatorEnodeTable struct {
+	valEnodeTable        map[common.Address]*validatorEnode
+	reverseValEnodeTable map[string]common.Address // enodeURL -> Address mapping
+	valEnodeTableMu      *sync.RWMutex             // This mutex protects both valEnodeTable and reverseValEnodeTable, since they are modified at the same time
+}
+
+// NewValidatorEnodeTable creates a validator-enodes-table
+func NewValidatorEnodeTable() *ValidatorEnodeTable {
+	vet := &ValidatorEnodeTable{
+		valEnodeTable:        make(map[common.Address]*validatorEnode),
+		reverseValEnodeTable: make(map[string]common.Address),
+		valEnodeTableMu:      new(sync.RWMutex),
+	}
+
+	return vet
+}
+
+func (vet *ValidatorEnodeTable) String() string {
+	var b strings.Builder
+	b.WriteString("ValEnodeTable:")
+	for _, valEnode := range vet.valEnodeTable {
+		fmt.Fprintf(&b, "%s\t", valEnode.String())
+	}
+	return b.String()
+}
+
+// GetEnodeURLFromAddress will return the enodeURL for a address if it's kwown
+func (vet *ValidatorEnodeTable) GetEnodeURLFromAddress(address common.Address) (string, bool) {
+	vet.valEnodeTableMu.RLock()
+	defer vet.valEnodeTableMu.RUnlock()
+	if enode, ok := vet.valEnodeTable[address]; ok {
+		return enode.enodeURL, ok
+	}
+	return "", false
+}
+
+// GetAddressFromEnodeURL will return the address for a enodeURL if it's kwown
+func (vet *ValidatorEnodeTable) GetAddressFromEnodeURL(enodeURL string) (common.Address, bool) {
+	vet.valEnodeTableMu.RLock()
+	defer vet.valEnodeTableMu.RUnlock()
+	address, ok := vet.reverseValEnodeTable[enodeURL]
+	return address, ok
+}
+
+// Upsert will update or insert a validator enode entry.  It will also do the associated set/remove to the validator connection.
+func (vet *ValidatorEnodeTable) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) (string, error) {
+	vet.valEnodeTableMu.Lock()
+	defer vet.valEnodeTableMu.Unlock()
+
+	var oldEnodeURL string
```

Comment:
I think `old` here is a bit misleading, since this entry may actually be more recent than the one we're trying to upsert. Wdyt of `current` or `existing`?